Source/Engine/Editor/VQAssetBrowser.h

```diff
@@ -0,0 +1,139 @@
+//	VQE
+//	Copyright(C) 2020  - Volkan Ilbeyli, Anilcan Gulkaya
+//
+//	This program is free software : you can redistribute it and / or modify
+//	it under the terms of the GNU General Public License as published by
+//	the Free Software Foundation, either version 3 of the License, or
+//	(at your option) any later version.
+//
+//	This program is distributed in the hope that it will be useful,
+//	but WITHOUT ANY WARRANTY; without even the implied warranty of
+//	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+//	GNU General Public License for more details.
+//
+//	You should have received a copy of the GNU General Public License
+//	along with this program.If not, see <http://www.gnu.org/licenses/>.
+//
+//	Contact: volkanilbeyli@gmail.com, anilcangulkaya7@gmail.com
+
+#pragma once
+#include "VQEditor.h"
+#include "../../Renderer/Renderer.h"
+
+namespace StringHash
+{
+	// very fast hash function + compile time
+	// http://www.cse.yorku.ca/~oz/hash.html
+	inline constexpr uint32 Basic(const char* str)
+	{
+		uint32 hash = 0; int c = 0;
+		while (c = *str++) hash += c;
+		return hash;
+	}
+}
+
+class AssetBrowser
+{
+public:
+	class FileRecord
+	{
+	public:
+		std::string path;
+		std::string name;
+		std::string extension;
+		CBV_SRV_UAV* texture = nullptr;
+		FileRecord() {}
+		FileRecord(std::string _path, std::string _name, std::string _extension)
+		:	path(std::move(_path)),
+			name(std::move(_name)),
+			extension(std::move(_extension)) {}
+	};
+
+	class FolderTree
+	{
+	public:
+		FolderTree() : parent(nullptr) {}
+		FolderTree(std::filesystem::path _path, std::string _name) : path(_path), name(std::move(_name)) {}
+	public:
+		std::filesystem::path  path;
+		FolderTree* parent = nullptr;
+		std::string name;
+		std::vector<FolderTree*> folders; // < -- subfolders
+		std::vector<FileRecord> files;
+	};
+
+	class TextureSRVPair
+	{
+	public:
+		TextureID textureid;
+		SRV_ID srv;
+		
+		CBV_SRV_UAV* GetTextureSRV(VQRenderer& renderer) const
+		{ 
+			return const_cast<CBV_SRV_UAV*>(&renderer.GetSRV(srv));
```

Comment:
Thanks for the links. I had to convert it to non const pointer because ImGui accepts non const pointer for ImGui::Image/ImageButton functions. You are using CBV_SRV_UAV* here:
![image](https://user-images.githubusercontent.com/48527900/170696543-cd16b0b0-4273-4d8f-bcbc-41715ea04e43.png)

Reply:
I think you need to provide non const version of GetSRV.

Reply:
No -- there is no need for a non-const GetSRV().

In fact, I should not be casting to `(CBV_SRV_UAV*)` here, `(const CBV_SRV_UAV*) would have been more appropriate.

Reply:
@benanil thanks for pointing the non-const pointer out -- fixed https://github.com/vilbeyli/VQEngine/commit/69dff0572aeda3f92fd51a875c92a552784bc42b